DbPDO::where change

gedtn opened the pull request:
There are a couple of changes here:

**Linux Changes**
 A couple of tweaks to get the dblib driver working properly (accessing authority from linux)

**DbPDO::where Changes**
Change the where method to handle the $equals arg being null. This change fixed a problem I was having with a query in the BI Refactor with multiple where-> clauses. Deep down in the fluent pdo code I found that each clause was generating a parameter, even when there was not one. This was causing an exception due to a mismatch in the number of arguments to the query. 

Originally I simply checked for a null $equals, but null can actually be passed as a parameter. Instead, I removed the "= null" in the argument list and checked for 2 args. 

I am a little wary as I would have expected that this should have been encountered already, and wondered, as this is a very recent machine (latest s/w and all..), that maybe it could be related to my envinronment? I checked and it does not seem that fluentPdo has changed in a while. 

I was not able to extensively test, as I had difficulties logging on via Linux. However, I was able to test:
- bi refactor code via a test action - accessing both local mysql and authority
- attendance employee synchronise
- intelligence extract

The change to the DbPDO::where is required for the BI Refactor to work, as coded. The other changes were required to allow me to test against authority locally. More extensive regression testing is required, which I would like to do on attendance test.

Should this also be applied to 0-8-0-BRANCH?

gedtn committed:
Fix where: do not pass null arg

gedtn committed:
Undo dev tweak

gedtn committed:
Use get_func_args to check for empty arg

adam-buckley commented:
DbPDO::where needs to be able to accept null! It is to clear all previous where clauses in the running statement.

See the documentation for FluentPDO [here](http://fluentpdo.com/documentation.html#where)

adam-buckley commented:
You also have some "echo" statements still in that code.

The line
```php
if (count(func_get_args()) == 2) {
```
is also redundant, as you've taken out the default null option for $equals, there will always be two arguments as PHP will stop if you don't give it two.

That's also why it's important to keep the null default for $equals in, as there may be places where we actually give null, and if there is (unless you've checked and there isn't) then this change could potentially break.

gedtn commented:
Agree with first point but I don't see where I have changed the interpretation of the first argument. Also, the echo was removed in a later commit. Where are you seeing it?

Regarding the core changes, I understand the crossed wires now. Please read the background section before considering the following:

I found, delving into the fluent dpo execution, that it can handle an empty second arg. To my mind, it makes sense to do so. 

Having a close look at the documentation was a good tip. If you go back up a few lines in the fluent pdo documentation to [here](http://fluentpdo.com/documentation.html#select), it lists:
```where($condition[, $parameters])```
My reading of that line is that $parameters are optional, assuming standard usage of square brackets. That would suggest that even though it is not explicitly listed as an example further down, having only the condition in the ```->where()``` is acceptable usage, and that we need to make these changes to DbPDO (pending further review).

**Background:**
Just to recap how I got here. When I ran the original code:
```
			$result = $connection->get("auhinode")->select("aupymast.emp_num, aupymast.nml_hrs, aupymast.nml_day")
					->leftJoin("aualclnk ON auhinode.nod_num = aualclnk.src_num")
					->leftJoin("aupymast ON aualclnk.alp_cd1 = aupymast.pos_cde")
					->where("auhinode.hcy_num", 6)
					->where("aualclnk.src_mdu", "HI")
					->where("aualclnk.lnk_typ", 60)
					->where("aupymast.tmd_dte ? OR aupymast.tmd_dte < getdate()", null)
					->where("aupymast.emp_num", $employee->employee_number)
					->orderBy("auhinode.nod_num");                        
```
I received the following exception:
>[Mon Mar 07 16:42:16.321668 2016] [:error] [pid 6124] [client 127.0.0.1:39022] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 156 General SQL Server error: Check messages from the SQL Server [156] (severity 15) [(null)]' in /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php:117\nStack trace:\n#0 /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php(117): PDOStatement->execute(Array)\n#1 /var/www/cmfive/wa-clean/system/lib/FluentPDO/SelectQuery.php(76): BaseQuery->execute()\n#2 /var/www/cmfive/wa-clean/system/classes/DbPDO.php(234): SelectQuery->fetch()\n#3 /var/www/bi-refactor/council/models/AuthorityService.php(73): DbPDO->fetch_row()\n#4 /var/www/bi-refactor/council/templates/test.tpl.php(43): AuthorityService->getEmployeeAuthorityRecord(Object(Employee))\n#5 /var/www/cmfive/wa-clean/system/web.php(1679): include('/var/www/bi-ref...')\n#6 /var/www/cmfive/wa-clean/system/web.php(1217): WebTemplate->fetch('modules/council...')\n#7 /var/www/cmfive/wa-clean/system/web.php(378): Web->fetchTemplate()\n#8 /var/www/cmfive/wa-clean/index.php(16): Web->star in /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php on line 117

I then extracted the sql, and found:
```sql
SELECT auhinode.*, aupymast.emp_num, aupymast.nml_hrs, aupymast.nml_day FROM auhinode LEFT JOIN aualclnk ON auhinode.nod_num = aualclnk.src_num LEFT JOIN aupymast ON aualclnk.alp_cd1 = aupymast.pos_cde WHERE auhinode.hcy_num = ? AND aualclnk.src_mdu = ? AND aualclnk.lnk_typ = ? AND aupymast.tmd_dte IS NULL OR aupymast.tmd_dte < getdate() AND aupymast.emp_num = ? ORDER BY auhinode.nod_num
```

I manually inserted the args and then ran via tsql, which returned many rows. I noticed that the where number 4 had an OR, but was coming through unbracketed. So I bracketed the clause and ran again, via tsql. I then received a nice single row that matched the query. I concluded there was nothing wrong with the query. 

Not knowing what the issue was, I tried putting the null on where number 4 inside the quotes, which is when I received the following error:
>[Mon Mar 07 16:43:01.915666 2016] [:error] [pid 9860] [client 127.0.0.1:39036] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY093]: Invalid parameter number: number of bound variables does not match number of tokens' in /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php:117\nStack trace:\n#0 /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php(117): PDOStatement->execute(Array)\n#1 /var/www/cmfive/wa-clean/system/lib/FluentPDO/SelectQuery.php(76): BaseQuery->execute()\n#2 /var/www/cmfive/wa-clean/system/classes/DbPDO.php(234): SelectQuery->fetch()\n#3 /var/www/bi-refactor/council/models/AuthorityService.php(73): DbPDO->fetch_row()\n#4 /var/www/bi-refactor/council/templates/test.tpl.php(43): AuthorityService->getEmployeeAuthorityRecord(Object(Employee))\n#5 /var/www/cmfive/wa-clean/system/web.php(1679): include('/var/www/bi-ref...')\n#6 /var/www/cmfive/wa-clean/system/web.php(1217): WebTemplate->fetch('modules/council...')\n#7 /var/www/cmfive/wa-clean/system/web.php(378): Web->fetchTemplate()\n#8 /var/www/cmfive/wa-clean/index.php(16): Web->start()\n#9 {main}\n  thrown in /var/www/cmfive/wa-clean/system/lib/FluentPDO/BaseQuery.php on line 117

It was at this point I delved in the fluent pdo execution. I was assuming that passing one 1 arg was acceptable. I found that it would accept an empty second arg, but DbPDO::where was always passing a second arg, giving null if none was given. This seemed to create the mismatch that lead to the second error.  

I tested changing DbPDO::where to pass through only one argument if only one was given (first cut was not correct, which is why I removed the ```=null``` and checked for the number of arguments). It worked.

gedtn commented:
Not sure where we are at. Any further thoughts/comments on this pull request?

adam-buckley commented:
I've checked this pull request and you still need to allow $equals to be null by default. I've tested this and it doesn't affect func_get_args count value.

This means that you can then call it with ```->where("X = Y")``` or ```->where("X", null)``` which is necessary to support all use cases.

Cheers,
Adam

gedtn commented:
Agreed....I will make this change, and do a final round of testing before committing.

gedtn committed:
Changes from pull request review

adam-buckley commented:
Hi Ged,

I just noticed that this pull request is still outstanding, is this still an issue? The code I'm looking at on master for the DbPDO::where() function looks ok to me... if this problem was fixed somewhere else let me know and I'll close this off.

gedtn commented:
Adam,

I think it is still an issue, but there is a work around.

Regards,
Ged

Regards,
Ged Nash


Mobile: 0427 883 282
ged@2pisoftware.com
http://2pisoftware.com


On Wed, Oct 12, 2016 at 1:54 PM, Adam Buckley <notifications@github.com>
wrote:

> Hi Ged,
>
> I just noticed that this pull request is still outstanding, is this still
> an issue? The code I'm looking at on master for the DbPDO::where() function
> looks ok to me... if this problem was fixed somewhere else let me know and
> I'll close this off.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/2pisoftware/cmfive/pull/194#issuecomment-253105455>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANZzdFOUI26BVSr5Y4GhGQjt1RV3yWSOks5qzEvNgaJpZM4HqfZT>
> .
>
